### PR TITLE
[WIP][Audit fix] Crab helper views

### PR DESCRIPTION
**Pull request by alpinechicken**
# Task:

## Description

Added view functions to CrabHelper to verify orders and check hedge size

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update

## How Has This Been Tested

New tests in test/integration-tests/crabv2/crab-helper.ts

## FE Checklist

- [x ] I have performed a self-review of my own code
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ x] My changes generate no new warnings
- [ ] Added video recordings if it is a UI change

## User Facing Checklist

- [ x] I fully understand the user problem this PR is solving
- [ x] I know who the target user is for this PR and have a deep understanding of that user
- [ x] I have tried this flow thinking from the pov of the target user for this PR
- [ ] I (or working w someone on team) have scheduled a user test for this PR (if it is a large change)

### CODE REVIEW

**Review by aparnakr — APPROVED** [review]

File: packages/hardhat/contracts/strategy/CrabHelper.sol [hunk 1]
```diff
@@ -97,6 +124,77 @@ contract CrabHelper is StrategySwap, ReentrancyGuard {
...
+        address offerSigner = ECDSA.recover(hash, _order.v, _order.r, _order.s);
+        require(offerSigner == _order.trader, "Invalid offer signature");
+        require(_order.expiry >= block.timestamp, "Order has expired");
+
+        // weth price for the order
+        uint256 wethAmount = _order.quantity.mul(_order.price).div(1e18);
+
+        if (_order.isBuying) {
+            // check weth balance and allowance
+            require(IWETH9(weth).balanceOf(_order.trader) >= wethAmount, "Not enough weth balance for trade");
```
[the remark this answers is not in the document]
Comment (aleone) [reply packages/hardhat/contracts/strategy/CrabHelper.sol:169]: Maybe return 3 booleans (isOrderBuyingoSQTH, isApproved, isBalanceSufficient).

If difficult/too late, can keep as is, no worries.

**Review by aleone — APPROVED** [review]
approving to not block but 2 main comments:
- maybe shouldn't have reverts in view functions and instead return what is wrong
- hedgingTwapPeriod shouldn't be declared in constructor, but pulled each time IMO as it can change

another general comment, can we check the view functions - can be just in the existing crab-hedging-otc file where we validate some correct orders, and check the failure cases (not enough weth, not enough oSQTH, approval, signature, etc)

File: packages/hardhat/contracts/strategy/CrabHelper.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (alpinechicken) [reply packages/hardhat/contracts/strategy/CrabHelper.sol:169]: should be easy to change. will go to (isValidNonce, isCorrectSignature, isNotExpired, isSufficientBalance, isSufficientAllowance)

File: packages/hardhat/contracts/strategy/CrabHelper.sol [hunk 2]
```diff
@@ -40,12 +64,15 @@ contract CrabHelper is StrategySwap, ReentrancyGuard {
...
 
-    constructor(address _crab, address _swapRouter) StrategySwap(_swapRouter) {
+    constructor(address _crab, address _swapRouter) StrategySwap(_swapRouter) EIP712("CrabOTC", "2") {
         require(_crab != address(0), "Invalid crab address");
 
         crab = _crab;
         weth = ICrabStrategyV2(_crab).weth();
         wPowerPerp = ICrabStrategyV2(_crab).wPowerPerp();
+        ethWSqueethPool = ICrabStrategyV2(_crab).ethWSqueethPool();
+        hedgingTwapPeriod = ICrabStrategyV2(_crab).hedgingTwapPeriod();
```
[the remark this answers is not in the document]
Comment (alpinechicken) [reply packages/hardhat/contracts/strategy/CrabHelper.sol]: the pool as well or just hedgingTwapPeriod?

File: packages/hardhat/contracts/strategy/CrabHelper.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (alpinechicken) [reply packages/hardhat/contracts/strategy/CrabHelper.sol:169]: should be easy to change. will go to (isValidNonce, isCorrectSignature, isNotExpired, isSufficientBalance, isSufficientAllowance)

File: packages/hardhat/contracts/strategy/CrabHelper.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (aleone) [reply packages/hardhat/contracts/strategy/CrabHelper.sol]: just the period, because the pool is always fixed

**Pull request closed without merging**